invokeai/app/services/shared/sqlite_migrator/sqlite_migrator_common.py

```diff
@@ -0,0 +1,209 @@
+    ) -> None:
+        self.name = name
+        self.dependency_type = dependency_type
+        self.value = None
+
+    def set_value(self, value: Any) -> None:
+        """
+        Sets the value of the dependency.
+        If the value is not of the correct type, a TypeError is raised.
+        """
+        if not isinstance(value, self.dependency_type):
+            raise TypeError(f"Dependency {self.name} must be of type {self.dependency_type}")
+        self.value = value
+
+
+class Migration(BaseModel):
+    """
+    Represents a migration for a SQLite database.
+
+    :param from_version: The database version on which this migration may be run
+    :param to_version: The database version that results from this migration
+    :param migrate: The callback to run to perform the migration
+    :param dependencies: A dict of dependencies that must be provided to the migration
+
+    Migration callbacks will be provided an open cursor to the database. They should not commit their
+    transaction; this is handled by the migrator.
+
+    Example Usage:
+    ```py
+    # Define the migrate callback. This migration adds a column to the sushi table.
+    def migrate_callback(cursor: sqlite3.Cursor, **kwargs) -> None:
+        # Execute SQL using the cursor, taking care to *not commit* a transaction
+        cursor.execute('ALTER TABLE sushi ADD COLUMN with_banana BOOLEAN DEFAULT TRUE;')
+        ...
+
+    # Instantiate the migration
+    migration = Migration(
+        from_version=0,
+        to_version=1,
+        migrate_callback=migrate_callback,
+    )
+    ```
+
+    If a migration needs an additional dependency, it must be provided with :meth:`provide_dependency`
+    before the migration is run. The migrator provides dependencies to the migrate callback,
+    raising an error if a dependency is missing or was provided the wrong type.
+
+    Example Usage:
+    ```py
+    # Create a migration dependency. This migration needs access the image files service, so we set the type to the ABC of that service.
+    image_files_dependency = MigrationDependency(name="image_files", dependency_type=ImageFileStorageBase)
+
+    # Define the migrate callback. The dependency may be accessed by name in the kwargs. The migrator will ensure that the dependency is of the required type.
+    def migrate_callback(cursor: sqlite3.Cursor, **kwargs) -> None:
+        image_files = kwargs[image_files_dependency.name]
+        # Do something with image_files
+        ...
+
+    # Instantiate the migration, including the dependency.
+    migration = Migration(
+        from_version=0,
+        to_version=1,
+        migrate_callback=migrate_callback,
+        dependencies={image_files_dependency.name: image_files_dependency},
+    )
+
+    # Provide the dependency before registering the migration.
+    # (DiskImageFileStorage is an implementation of ImageFileStorageBase)
+    migration.provide_dependency(name="image_files", value=DiskImageFileStorage())
```

Comment:
Here's an alternative solution to the migration dependency problem. Take it or leave it. Your implementation is already in good shape (and you've spent more time thinking about this than me), so I'll defer to you to make the call. But, I'll give a couple reasons why I think this is better.

```python
# ----- migrate_2.py -----
class Migration2Function:
    def __init__(self, image_files: ImageFileStorageBase, logger: Logger):
        self._image_files = image_files
        self._logger = logger

    def __call__(self, cursor: sqlite3.Cursor):
        # Run the migration with access to the self._image_files and self._logger dependencies.
        ...

def build_migration_2(image_files: ImageFileStorageBase, logger: Logger) -> Migration:
    # Could alternatively use `partial(...)` here if you prefer that over using a class.
    migration_fn = Migration2Function(image_files, logger)

    return Migration(
        from_version=1,
        to_version=2,
        migrate_callback=migration_fn,
    )

# ----- invokeai/app/api/dependencies.py -----
...

migrator = SQLiteMigrator(db=db)
migrator.register_migration(build_migration_1())
migrator.register_migration(build_migration_2(image_files, logger))
migrator.run_migrations()
...

```

Advantages of this approach:
- Simpler. Can drop all of the`MigrationDependency` handling.
- We don't have to worry about runtime dependency type checking. We benefit from static type checking (or at least as much as we ever do in python).
- No code gets run on import of the migration files. I.e. when we import migration_1.py, we don't instantiate a migration, we just declare classes/functions. I try to avoid running code on import, because it can lead to confusing bugs where import order matters.

Reply:
Oh, this is much nicer. I was conceptually orbiting something like this, but ended up with the contrived dependency handling. I can't think of any reason to use my implementation of the dependencies. I'll do this instead, thanks!

Using a class also allows the smaller logical migration functions to be local to the class - having those in the module scope was bothering me.

Also cool that you can just make a class callable with `__call__`. I knew you could make protocols like this but it never clicked.